### PR TITLE
added the possibility of merging bams from a parallelized job

### PR DESCRIPTION
**Pull request by roalva1**
We are trying to contribute to the Galaxy parallelization module, so we add the merge static method to the Bam class.

### CODE REVIEW

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -169,6 +169,11 @@ def _get_samtools_version( self ):
                 break
         return version
 
+    @staticmethod
+    def merge(split_files, output_file):
+        params = ["samtools", "merge", "-f", output_file, " ".join(split_files)]
```
Comment (jmchilton) [inline lib/galaxy/datatypes/binary.py]: Can you place the split_files in the `params`  array and disable `shell=True`.

Comment (peterjc) [inline lib/galaxy/datatypes/binary.py]: I was wondering about `shell=True`, ideally from a security point none of the existing calls in `binary.py` would do that either.

Comment (roalva1) [inline lib/galaxy/datatypes/binary.py]: What do you mean exactly with "place the split_files in the params"? Do you mean to change
params = ["samtools", "merge", "-f", output_file, " ".join(split_files)]
by
params = ["samtools", "merge", "-f", output_file, split_files]

and what about shell? do you recommend to leave it as True or False?

Comment (jmchilton) [inline lib/galaxy/datatypes/binary.py]: I assume split_files is a list, if yes then something like this should work:

params = ["samtools", "merge", "-f", output_file] + split_files

I would then set `shell=False` or remove the parameter and should be the default.

**Comment by roalva1** [comment]
Finally, I write the code as it is in the rest of the functions in the same file, is it ok?

**Comment by peterjc** [comment]
@jmchilton Are there any tests specifically covering the split/merge functionality?

**Comment by roalva1** [comment]
What I do to try the code is to use some human fastq that I have and then put this line in the bwa tool devteam repo (shed_tools/toolshed.g2.bx.psu.edu/repos/devteam/bwa/c71dd035971e/bwa/bwa-mem.xml):

parallelism method="multi" split_size="3" shared_inputs="ref_file" split_mode="number_of_parts" merge_outputs="bam_output" split_inputs="fastq_input1,fastq_input2" ></parallelism

**Comment by jmchilton** [comment]
:+1: Thanks for the contribution and making the requested fixes. I have some WIP testing against this branch downstream that adds an automated test confirming this works. https://github.com/galaxyproject/galaxy/commit/06a3f272e2eba80713124ecb6d37594e7b313f68

**Comment by roalva1** [comment]
Hello, I am trying to get that commit in order to try by myself these tests but I am not able to find it. I have done a fetch and then looking for the commit: 06a3f272e2eba80713124ecb6d37594e7b313f68, but nothing, it is not in the tree, would you tell me if there is anyway of getting this code?

**Comment by jmchilton** [comment]
@roalva1 Sorry about that - github did something unwanted with the link there. I have wrapped a new cleaner link in bitly - http://bit.ly/1JS2IQC. The test case is no longer passing for me - I don't know what happened.

**Comment by roalva1** [comment]
As the code and the test is in the dev branch I can try the test and see what happening

**Comment by roalva1** [comment]
I have been looking for the problem, and it could be in the tool in this line:
samtools view $input1 | grep \$filter | samtools view -b - > $out

I am trying to run this command directly, for example:

<pre>samtools view  test-data/1.bam | grep 10864 | samtools view -b -</pre>

and this is output:

<pre>[bam_header_read] EOF marker is absent. The input is probably truncated.
[bam_header_read] invalid BAM binary header (this is not a BAM file).
[main_samview] fail to read the header from "-".
</pre>

I guess is because the grep command is filtering the header, so at the end, there is no bams to merge

**Pull request merged**